pywbem_mock/_methodprovider.py

```diff
@@ -180,86 +180,114 @@ def __init__(self, cimrepository=None):
-            The object path of the target object, as follows:
+          objectname: (:class:`~pywbem.CIMInstanceName` or :class:`~pywbem.CIMClassName`):
+            A reference to the target CIM object, as follows:
 
             * For instance-level use: The instance path of the target
-              instance, as a :class:`~pywbem.CIMInstanceName` object.
-              If this object does not specify a namespace, the default namespace
-              of the connection is used.
-              Its `namespace`, and `host` attributes will be ignored.
-
-            * For class-level use: The class name of the target class, as a
-              :term:`string`. This should be one of the classes defined in
-              the class variable `classnames`.
-
-              The string is interpreted as a class name in the namespace defined
-              in the namespace parameter(case independent).
+              instance, as a :class:`~pywbem.CIMInstanceName` object, with the
+              following attributes:
+              - `classname`: Will be set, in any lexical case.
+              - `keybindings`: Will be set, with key names in any lexical case.
+              - `namespace`: Will be set, in any lexical case, and with leading
+                and trailing slash characters removed.
+              - `host`: Will be `None`.
+
+            * For class-level use: The class path of the target class, as a
+              :class:`~pywbem.CIMClassName` object, with the following
+              attributes:
+              - `classname`: Will be set, in any lexical case.
+              - `namespace`: Will be set, in any lexical case, and with leading
+                and trailing slash characters removed.
+              - `host`: Will be `None`.
+
+          params (:class:`py:NocaseDict`):
+            The input parameters for the method invocation, with items as
+            follows:
+            - key (:term:`string`): The parameter name, in any lexical case.
+            - value (:class:`~pywbem.CIMParameter`): The parameter value.
 
-          Params (:class:`py:NocaseDict`):
-            Each item in ``Params`` is a name/parameter-value for the CIM
-            method and is:
+        Returns:
 
-            * :term:`string`: The case-insensitive name of the parameter.
-            * :class:`~pywbem.CIMParameter`: The item value representing a
-              parameter value. The `name`, `value`, `type` and
-              `embedded_object` attributes of this object are guaranteed
-              present.
+            :func:`py:tuple` (return_value, out_params): The return value and
+            output parameters of the method invocation:
 
-        Returns:
+            * return_value (:term:`CIM data type`):
+              Return value of the method invocation.
 
-            A :func:`py:tuple` of (returnvalue, outparams), with these
-            tuple items:
+            * out_params (list/tuple or dict):
```

Comment:
Why have multiple ways to construct the output parameters,  I would thing that just allowing CIMParameter would reduce the number of alternatives for the user.

Reply:
The client operation does not return CIMParameter, but only the value (as dict(name:value)). So requiring the detour through CIMParameter is overhead, but since it is quite convenient I did not want to prohibit it.

On input parameters, the client operation is flexible to accept a list/tuple of CIMParameter and dict(name:value), which is transformed to only list/tuple of CIMParameter as input to the provider method. So there we already have the precedent case that the user has some flexibility in what they provide, and we narrow it down to a single representation.

The output parameter apply this principle as well, just this time the user provides the output parameter in the provider method (with flexibility), and we narrow it down to a single representation when passing it out of the client operation.

So there is some logic and consistency behind that.

Having said that, I'm open for discussion.

DISCUSSION

Reply:
Karl agrees to the approach.